cpp/cuproj/include/cuproj/operation/transverse_mercator.cuh

```diff
@@ -0,0 +1,518 @@
+      n * (0.5 + n * (-2 / 3.0 + n * (5 / 16.0 + n * (41 / 180.0 +
+                                                      n * (-127 / 288.0 + n * (7891 / 37800.0))))));
+    tmerc_params.utg[1] =
+      np * (-1 / 48.0 +
+            n * (-1 / 15.0 + n * (437 / 1440.0 + n * (-46 / 105.0 + n * (1118711 / 3870720.0)))));
+    tmerc_params.gtu[1] =
+      np * (13 / 48.0 +
+            n * (-3 / 5.0 + n * (557 / 1440.0 + n * (281 / 630.0 + n * (-1983433 / 1935360.0)))));
+    np *= n;
+    tmerc_params.utg[2] =
+      np * (-17 / 480.0 + n * (37 / 840.0 + n * (209 / 4480.0 + n * (-5569 / 90720.0))));
+    tmerc_params.gtu[2] =
+      np * (61 / 240.0 + n * (-103 / 140.0 + n * (15061 / 26880.0 + n * (167603 / 181440.0))));
+    np *= n;
+    tmerc_params.utg[3] = np * (-4397 / 161280.0 + n * (11 / 504.0 + n * (830251 / 7257600.0)));
+    tmerc_params.gtu[3] = np * (49561 / 161280.0 + n * (-179 / 168.0 + n * (6601661 / 7257600.0)));
+    np *= n;
+    tmerc_params.utg[4] = np * (-4583 / 161280.0 + n * (108847 / 3991680.0));
+    tmerc_params.gtu[4] = np * (34729 / 80640.0 + n * (-3418889 / 1995840.0));
+    np *= n;
+    tmerc_params.utg[5] = np * (-20648693 / 638668800.0);
+    tmerc_params.gtu[5] = np * (212378941 / 319334400.0);
+
+    // Gaussian latitude value of the origin latitude
+    T const Z = detail::gatg(
+      tmerc_params.cbg, ETMERC_ORDER, params.phi0, cos(2 * params.phi0), sin(2 * params.phi0));
+
+    // Origin northing minus true northing at the origin latitude
+    // i.e. true northing = N - P->Zb
+    tmerc_params.Zb =
+      -tmerc_params.Qn * (Z + detail::clenshaw_real(tmerc_params.gtu, ETMERC_ORDER, 2 * Z));
+
+    return params;
+  }
+
+ private:
+  /**
+   * @brief  Forward projection, from geographic to transverse mercator.
+   *
+   * @param coord Geographic coordinate (lat, lon) in radians.
+   * @return Transverse mercator coordinate (x, y) in meters.
+   */
+  __host__ __device__ Coordinate forward(Coordinate const& coord) const
+  {
+    // so we don't have to qualify the class name everywhere.
+    auto& tmerc_params = this->params_.tmerc_params_;
+    auto& ellipsoid    = this->params_.ellipsoid_;
+
+    // ell. LAT, LNG -> Gaussian LAT, LNG
+    T Cn =
+      detail::gatg(tmerc_params.cbg, ETMERC_ORDER, coord.y, cos(2 * coord.y), sin(2 * coord.y));
+
+    // Gaussian LAT, LNG -> compl. sph. LAT
+    T const sin_Cn = sin(Cn);
+    T const cos_Cn = cos(Cn);
+    T const sin_Ce = sin(coord.x);
+    T const cos_Ce = cos(coord.x);
+
+    T const cos_Cn_cos_Ce = cos_Cn * cos_Ce;
+    Cn                    = atan2(sin_Cn, cos_Cn_cos_Ce);
+
+    T const inv_denom_tan_Ce = 1. / hypot(sin_Cn, cos_Cn_cos_Ce);
+    T const tan_Ce           = sin_Ce * cos_Cn * inv_denom_tan_Ce;
+
+#if 0
+    // Variant of the above: found not to be measurably faster
+    T const sin_Ce_cos_Cn = sin_Ce*cos_Cn;
+    T const denom = sqrt(1 - sin_Ce_cos_Cn * sin_Ce_cos_Cn);
+    T const tan_Ce = sin_Ce_cos_Cn / denom;
+#endif
```

Comment:
Commenting it out is better than using a macro to disable it.

Reply:
Yep, copied from PROJ like this

Reply:
Fixed